pkg/distributor/distributor.go

```diff
@@ -309,6 +313,41 @@ func (d *Distributor) PushParsed(ctx context.Context, req *distributormodel.Push
 		)
 	}
 
+	// Normalisation is quite an expensive operation,
+	// therefore it should be done after the rate limit check.
+	for _, series := range req.Series {
+		for _, sample := range series.Samples {
+			sample.Profile.Normalize()
+		}
+	}
+
+	if len(req.Series) == 1 && len(req.Series[0].Samples) == 1 {
+		series := req.Series[0]
+		// Actually all series profiles can be merged before aggregation.
+		// However, it's not expected that a series has more than one profile.
+		profile := series.Samples[0].Profile.Profile
+		var aggregated aggregator.AggregationResult[*pprof.ProfileMerge]
+		if aggregated, err = d.maybeAggregate(tenantID, series.Labels, profile); err != nil {
```

Comment:
Distributors are randomly assigned series of profiles, so I think this affect the effectiveness of aggregating profiles ?

Reply:
>  Note that the period determines the lowest latency of the ingestion requests

I'm also trying to understand something, are you holding all pusher until the aggregated value is pushed ?

Reply:
> Distributors are randomly assigned series of profiles, so I think this affect the effectiveness of aggregating profiles ?

That's very true. Therefore aggregation only makes sense with low cardinality data. Moreover, `max_sessions_per_series` should be finely tuned, if enabled. Otherwise, aggregation may not trigger. 

> I'm also trying to understand something, are you holding all pusher until the aggregated value is pushed ?

All queries contributing to the aggregate are blocked until it is processed. This is the only way to guarantee at-least-once delivery: if we fail to send the unit to the receivers, we need to pass the error to the clients so they can handle it accordingly (e.g., try again). However, there is one danger: if a failure occurs on the client side after processing the aggregate and it repeats the sending, this will lead to duplication of data. For a serverless scenario this isn't as important as they should not (or can't) retry.

Now, as I write this, I wonder: If we tell clients not to try again, what's the point of blocking them? Perhaps we should release them immediately if we can guarantee they won't try again.